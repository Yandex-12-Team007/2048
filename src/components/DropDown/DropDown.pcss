.drop-down {
  width: 100%;
  height: 100%;
  position: relative;

  &__label {
    width: 100%;
    height: 100%;

    display: flex;
    justify-content: center;
    align-items: center;
  }

  &__menu {
    z-index: 1000;
    position: absolute;
    top: 100%;
    right: 0;

    box-sizing: border-box;
    margin: 5px 0 0 0;
    padding: 10px;

    min-width: 160px;

    background: var(--second-bg);

    display: flex;
    justify-content: center;
    align-items: center;
    flex-direction: column;

    border-radius: 15px;
  }
}

.drop-down-item {
  width: 100%;
  height: 32px;
  margin: 5px;

  text-align: center;

  &__btn {
    color : #000;

    &:active,
    &:hover {
      color: inherit;
    }
  }

  &__link {
    color : #000;
    text-decoration: none;

    &:active,
    &:hover {
      text-decoration: none;
      color: inherit;
    }
  }
}
