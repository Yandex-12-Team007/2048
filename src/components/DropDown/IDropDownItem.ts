export enum IDropDownItemType {
  BUTTON,
  LINK
}

export interface IDropDownItem {
  type : IDropDownItemType,
  title : string,
  action? : () => void,
  link? : string
}
