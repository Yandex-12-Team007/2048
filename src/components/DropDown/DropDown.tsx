import React, {useState} from 'react';
import {Link} from 'react-router-dom';

import {IDropDownItem, DropDownItemType} from './IDropDownItem';

import './DropDown.pcss';

export default function DropDown({children, options} : {
  children,
  options : IDropDownItem[]
}) {
  const [isOpen, setOpen] = useState(false);

  return <div className={'drop-down'}>
    <div className={'drop-down__label'} onClick={() => setOpen(!isOpen)}>
      {children}
    </div>
    {isOpen ?
      <div className={'drop-down__menu'}>
        {options.map((el, id) => <DropDownItem key={id} item={el}/>)}
      </div> : ''
    }
  </div>
}

// TODO: Возможно стоит сделать прокидываемым
function DropDownItem({item} : {item : IDropDownItem}) {
  if (item.type === DropDownItemType.BUTTON) {
    return <div className={'drop-down-item'}>
      <button className={'drop-down-item__btn'} onClick={item.action}>
        {item.title}
      </button>
    </div>
  }

  if (item.type === DropDownItemType.LINK) {
    return <div className={'drop-down-item'}>
      <Link className={'drop-down-item__link'} to={item.link}>
        {item.title}
      </Link>
    </div>
  }

  /* Если попадается неизвестный тип - исключение */
  throw new Error('Undefined drop down item type !');
}
