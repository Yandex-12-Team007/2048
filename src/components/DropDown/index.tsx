import DropDown from './DropDown';
import {IDropDownItem, DropDownItemType} from './IDropDownItem';

export {
  IDropDownItem,
  DropDownItemType,
};

export default DropDown;
