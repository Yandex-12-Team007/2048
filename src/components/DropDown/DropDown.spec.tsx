import React from 'react';
import renderer from 'react-test-renderer';
import DropDown, {DropDownItemType} from './';

it('Dropdown with options', () => {
  const options = [
    {
      title: 'test link',
      type: DropDownItemType.LINK,
      link: '#',
    },
    {
      title: 'test action',
      type: DropDownItemType.BUTTON,
      action: () => console.log('test'),
    },
  ];

  const tree = renderer.create(
      <DropDown options={options} >
        <h2>Test</h2>
      </DropDown>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
