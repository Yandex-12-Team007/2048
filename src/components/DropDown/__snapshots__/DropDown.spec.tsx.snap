// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown with options 1`] = `
<div
  className="drop-down"
>
  <div
    className="drop-down__label"
    onClick={[Function]}
  >
    <h2>
      Test
    </h2>
  </div>
</div>
`;
